src/main/java/org/prebid/server/bidder/huaweiads/HuaweiAdsBidder.java

```diff
@@ -0,0 +1,1267 @@
+                    && !openRTBRequest.getApp().getContent().getLanguage().isEmpty()) {
+
+                app.lang(openRTBRequest.getApp().getContent().getLanguage());
+            } else {
+                app.lang("en");
+            }
+        }
+        final String countryCode = getCountryCode(openRTBRequest);
+        app.country(countryCode);
+        return app.build();
+    }
+
+    private String getFinalPackageName(String bundleName) {
+        for (PkgNameConvert convert : packageNameConverter) {
+            if (convert.getConvertedPkgName().isEmpty()) {
+                continue;
+            }
+
+            for (String name : convert.getExceptionPkgNames()) {
+                if (name.equals(bundleName)) {
+                    return bundleName;
+                }
+            }
+
+            for (String name : convert.getUnconvertedPkgNames()) {
+                if (name.equals(bundleName) || name.equals("*")) {
+                    return convert.getConvertedPkgName();
+                }
+            }
+
+            for (String keyword : convert.getUnconvertedPkgNameKeyWords()) {
+                if (bundleName.indexOf(keyword) > 0) {
+                    return convert.getConvertedPkgName();
+                }
+            }
+
+            for (String prefix : convert.getUnconvertedPkgNamePrefixs()) {
+                if (bundleName.startsWith(prefix)) {
+                    return convert.getConvertedPkgName();
+                }
+            }
+        }
+
+        return bundleName;
+    }
+
+    private Device getDeviceInfo(BidRequest openRTBRequest) {
+        final Device.DeviceBuilder deviceBuilder = Device.builder();
+        Optional.ofNullable(openRTBRequest.getDevice())
+                .ifPresent(deviceData -> {
+                    deviceBuilder.type(deviceData.getDevicetype());
+                    deviceBuilder.useragent(deviceData.getUa());
+                    deviceBuilder.os(deviceData.getOs());
+                    deviceBuilder.version(deviceData.getOsv());
+                    deviceBuilder.maker(deviceData.getMake());
+                    deviceBuilder.model(StringUtils.isEmpty(deviceData.getModel())
+                            ? HuaweiAdsConstants.DEFAULT_MODEL_NAME : deviceData.getModel());
+                    deviceBuilder.height(deviceData.getH());
+                    deviceBuilder.width(deviceData.getW());
+                    deviceBuilder.language(deviceData.getLanguage());
+                    deviceBuilder.pxratio(deviceData.getPxratio());
+                    String countryCode = getCountryCode(openRTBRequest);
+                    deviceBuilder.belongCountry(countryCode);
+                    deviceBuilder.localeCountry(countryCode);
+                    deviceBuilder.ip(deviceData.getIp());
+                    deviceBuilder.gaid(deviceData.getIfa());
+                });
+
+        return getDeviceIdFromUserExt(deviceBuilder.build(), openRTBRequest);
```

Comment:
So `DeviceId` or `Device`? Because method return type is actually `Device` and not `String` or some form of ID type